### PR TITLE
support segment split by scanner

### PR DESCRIPTION
**Pull request by HappenLee**
# Proposed changes

Issue Number: close #xxx

## Problem summary

Describe your changes.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by HappenLee** [comment]
run buildall

**Comment by HappenLee** [comment]
run buildall

**Comment by HappenLee** [comment]
run buildall

**Comment by HappenLee** [comment]
run buildall

**Comment by HappenLee** [comment]
run buildall

File: build.sh [hunk 1]
```diff
@@ -123,7 +123,7 @@ fi
 
 eval set -- "${OPTS}"
 
-PARALLEL="$(($(nproc) / 4 + 1))"
+PARALLEL="$(($(nproc) / 2 + 1))"
```
Comment (wangbo) [inline build.sh:126]: Why change it?

**Comment by HappenLee** [comment]
run buildall

File: be/src/olap/rowset/beta_rowset_reader.cpp [hunk 2]
```diff
@@ -174,7 +179,15 @@ Status BetaRowsetReader::get_segment_iterators(RowsetReaderContext* read_context
                                                            should_use_cache));
 
     // create iterator for each segment
-    for (auto& seg_ptr : _segment_cache_handle.get_segments()) {
+    auto& segments = _segment_cache_handle.get_segments();
+    auto [seg_start, seg_end] = segment_offset;
+    if (seg_start == seg_end) {
```
Comment (wangbo) [inline be/src/olap/rowset/beta_rowset_reader.cpp:184]: Is there any possible that current rowset reader is allocated [seg_start=10,seg_end=10], then seg_start == seg_end, seg_start and seg_end is reset to [0, segments.size()]

**Comment by wangbo** [comment]
Please add some description information for this PR

File: be/src/vec/exec/scan/new_olap_scanner.cpp [hunk 3]
```diff
@@ -72,12 +74,6 @@ Status NewOlapScanner::prepare(const TPaloScanRange& scan_range,
     {
         std::string err;
         _tablet = StorageEngine::instance()->tablet_manager()->get_tablet(tablet_id, true, &err);
-        if (_tablet.get() == nullptr) {
```
Comment (wangbo) [inline be/src/vec/exec/scan/new_olap_scanner.cpp:75]: Why remove tablet nullptr check here?

File: be/src/vec/exec/scan/new_olap_scan_node.cpp [hunk 4]
```diff
@@ -404,57 +406,176 @@ Status NewOlapScanNode::_init_scanners(std::list<VScanner*>* scanners) {
...
+            if (!rs_readers.empty()) {
+                build_new_scanner(*scan_range, scanner_ranges, rs_readers, rs_reader_seg_offsets);
+            }
         }
+    } else {
+        for (auto& scan_range : _scan_ranges) {
+            auto tablet_id = scan_range->tablet_id;
+            std::string err;
+            TabletSharedPtr tablet =
+                    StorageEngine::instance()->tablet_manager()->get_tablet(tablet_id, true, &err);
```
Comment (wangbo) [inline be/src/vec/exec/scan/new_olap_scan_node.cpp:550]: Missing tablet nullptr check here

File: be/src/olap/rowset/beta_rowset_reader.cpp (hunk 2 above)
Comment (HappenLee) [reply be/src/olap/rowset/beta_rowset_reader.cpp:184]: No. if spilit, unless rowset segment is empty. the segment_offset must have one segment by split

**Comment by HappenLee** [comment]
run buildall

**Review by wangbo — APPROVED** [review]
LGTM

**Comment by wangbo** [comment]
https://github.com/apache/doris/issues/17852

**Review by yiguolei — APPROVED** [review]
LGTM

**Pull request merged**